docs/upgrade.html

```diff
@@ -17,6 +17,52 @@
 
+<h4><a id="upgrade_10" href="#upgrade_10">Upgrading from 0.9.x to 0.10.0.0</a></h4>
+0.10.0.0 has <a href="#upgrade_10_performance_impact">potential performance impact during upgrade</a> and <a href="#upgrade_10_protocol_changes">protocol changes</a> (please review before upgrading).
+Because new protocols are introduced, it is important to upgrade your Kafka clusters before upgrading your clients.
+
+<p><b>For a rolling upgrade:</b></p>
+
+<ol>
+    <li> Update server.properties file on all brokers and add the following properties: inter.broker.protocol.version=0.9.0.0, message.format.version=v0 </li>
+    <li> Upgrade the brokers. This can be done a broker at a time by simply bringing it down, updating the code, and restarting it. </li>
+    <li> Once the entire cluster is upgraded, bump the protocol version by editing inter.broker.protocol.version and setting it to 0.10.0.0. </li>
+    <li> Restart the brokers one by one for the new protocol version to take effect. </li>
+</ol>
+
+<p><b>Note:</b> If you are willing to accept downtime, you can simply take all the brokers down, update the code and start all of them. They will start with the new protocol by default.
+
+<p><b>Note:</b> Bumping the protocol version and restarting can be done any time after the brokers were upgraded. It does not have to be immediately after.
+
+<h5><a id="upgrade_10_performance_impact" href="#upgrade_10_performance_impact">potential performance impact in 0.10.0.0 during upgrade</a></h5>
+<p>
+    Message format v1 is added to include a new timestamp field and use relative offsets for compressed messages in 0.10.0.0.
+    The on disk message format can be configured through message.format.version in server.properties file.
+    The default on-disk message format in 0.10.0 in v1. If a consumer client is on a version before 0.10.0.0, it only understands
+    message format v0. In this case, the broker is able to convert messages of format v1 to v0 before sending a response
+    to the consumer on an older version. However, the broker can't use zero-copy transfer in this case.
+
+    To avoid such message conversion before consumers are upgraded to 0.10.0, one can set the message format to v0 after
+    upgrading the broker to 0.10.0. This way, the broker can still use zero-copy transfer to send the data to the old
+    consumers. Once most consumers are upgraded, one can change the message format to v1 on the broker.
+</p>
+<p>
+    0.10.0.0 clients can handle both message format v0 and v1. Hence there is no performance impact for 0.10.0.0 clients
+    if message.format.version is set to v0 on broker. Broker will not convert the messages but just return message v0
+    to 0.10.0.0 consumer.
+</p>
+<p>
+    <b>Note:</b> By setting the message format version, one certifies all the existing messages is on or below that
+    message format version. Otherwise consumers before 0.10.0.0 might break.
```

Comment:
Could we also add sth like the following.

In particular, after the message format is set to v1, one should not change it back to v0 since it may break the consumer on versions before 0.10.0.